src/Rewards.sol

```diff
@@ -0,0 +1,239 @@
+
+        if (block.timestamp >= quotaRenewalTimestamp) {
+            rewardsClaimed = 0;
+
+            // The following operations are safe based on the constructor's requirements for longer than the age of universe :)
+            uint256 timeAhead = block.timestamp - quotaRenewalTimestamp;
+            quotaRenewalTimestamp = block.timestamp + rewardPeriod - (timeAhead % rewardPeriod);
+        }
+
+        (bool sucess, uint256 newRewardsClaimed) = rewardsClaimed.tryAdd(reward.amount);
+        if (!sucess || newRewardsClaimed > rewardQuota) {
+            revert RewardQuotaExceeded();
+        }
+        rewardsClaimed = newRewardsClaimed;
+
+        // Safe to increment the sequence after checking this is the expected number (no overflow for the age of universe even with 1000 reward claims per second)
+        rewardSequences[reward.recipient] = reward.sequence + 1;
+
+        nodlToken.mint(reward.recipient, reward.amount);
+
+        emit RewardMinted(reward.recipient, reward.amount, rewardsClaimed);
+    }
+
+    function _mustBeGreaterThanZero(uint256 value) internal pure {
+        if (value == 0) {
+            revert ZeroPeriod();
+        }
+    }
+    /**
+     * @dev Internal check to ensure the `sequence` value is expected for `receipent`.
+     * @param receipent The address of the receipent to check.
+     * @param sequence The sequence value.
+     */
+
+    function _mustBeExpectedSequence(address receipent, uint256 sequence) internal view {
+        if (rewardSequences[receipent] != sequence) {
+            revert InvalidRecipientSequence();
+        }
+    }
+
+    /**
+     * @dev Checks if the provided signature is valid for the given hash and authorized oracle address.
+     * @param hash The hash to be verified.
+     * @param signature The signature to be checked.
+     * @dev Throws an `UnauthorizedOracle` exception if the signature is not valid.
+     */
+    function _mustBeFromAuthorizedOracle(bytes32 hash, bytes memory signature) internal view {
+        if (!SignatureChecker.isValidSignatureNow(authorizedOracle, hash, signature)) {
+            revert UnauthorizedOracle();
+        }
+    }
+
+    /**
+     * @dev Helper function to get the digest of the typed data to be signed.
+     * @param reward detailing recipient, amount, and sequence.
+     * @return The hash of the typed data.
+     */
+    function digestReward(Reward memory reward) external view returns (bytes32) {
+        return digest(reward);
+    }
+
+    /**
+     * @dev Internal helper function to get the digest of the typed data to be signed.
+     * @param reward detailing recipient, amount, and sequence.
+     * @return The hash of the typed data.
+     */
+    function digest(Reward memory reward) internal view returns (bytes32) {
+        return _hashTypedDataV4(
+            keccak256(abi.encode(keccak256(REWARD_TYPE), reward.recipient, reward.amount, reward.sequence))
+        );
+    }
```

Comment:
I am not super clear why we have these two functions instead of just one. Couldn't we do something like the below?
```suggestion
    /**
     * @dev Internal helper function to get the digest of the typed data to be signed.
     * @param reward detailing recipient, amount, and sequence.
     * @return The hash of the typed data.
     */
    function digestReward(Reward memory reward) public view returns (bytes32) {
        return _hashTypedDataV4(
            keccak256(abi.encode(keccak256(REWARD_TYPE), reward.recipient, reward.amount, reward.sequence))
        );
    }
```

Reply:
I had tried it before. But now I realized when I tested it, the function was marked as external while it should have been public

Reply:
According to the book that you shared with me, my previous approach could have been more gas efficient: "Use external functions: Functions should be marked as external whenever possible, as external functions are more gas-efficient than public ones. This is because external functions expect arguments to be passed from the external call, which can save gas."

Reply:
Yes that would be correct indeed